### PR TITLE
Memberships: Add coupons GUI

### PR DESCRIPTION
**Pull request by chrismccluskey**
Related to 225-gh-automattic/gold

## Proposed Changes

* Adding a coupons GUI for memberships.

## Translators

You should be able to find each of the strings in these screenshots

### Coupons list

- `%s off for 1 month`
- `%s off for 3 months`
- `%s off for 6 months`
- `%s off for 1 year`
- `%s off forever`
- `Manage coupons`
- `Add a new coupon`
- `First-time order only`
- `Limited to specific emails`
- `Limited to specific products`

Screenshot 2023-12-29 at 12 52 09 PM

#### Any product
- `Any product`
- `Choose a unique coupon code for the discount. Not case-sensitive.`
- `Percentage`
- `Expiration date (optional)`
- `Usage limit (optional)`
- `Limit the number of times this coupon can be redeemed by a supporter.`
- `Eligible for first-time order only`
- `Once the coupon is used for a subscription, how long does it last?`
- `Limit coupon to specific emails`
- `Separate email addresses with commas. Use an asterisk (*) to match parts of an email. For example, "*@university.edu" would select all "university.edu" addresses.`

Screenshot 2023-12-29 at 12 52 22 PM

#### 1 Product Selected
- `1 product selected`

Screenshot 2023-12-29 at 12 52 55 PM

#### Multiple products selected
- `%s products selected.`

Screenshot 2023-12-29 at 12 53 16 PM

#### Duration
- `3 Months`
- `6 Months`
- `1 Year`

Screenshot 2023-12-29 at 1 04 55 PM

#### Errors
NOTE: To get these error message, click or tab into the input field, then tab or click away without entering any information, or by entering purposefully invalid information (negative numbers, end dates before start dates, etc..)

##### Invalid coupon code
- `Please input a coupon code.`

Screenshot 2023-12-29 at 1 06 16 PM

##### Invalid type
This error is difficult to trigger, which is why there is no screenshot. However, it will look like the other errors and will be under the "Discount type" selector.
- `Please select a valid type`

##### Invalid amount
- `Please enter a valid amount`

Screenshot 2023-12-29 at 1 13 52 PM

##### Invalid dates
- `Please enter a valid start date`
- `Please enter a valid end date`

Screenshot 2023-12-29 at 1 14 59 PM

##### Invalid usage limit
- `Please enter a positive value, or 0 for infinite uses.`

Screenshot 2023-12-29 at 1 15 42 PM

##### Invalid emails
- `Please fix these invalid email addresses: %s`

Screenshot 2023-12-29 at 1 16 06 PM

### Delete coupon?
- `Delete coupon?`
- `Deleting this coupon ({{strong}}%s{{/strong}}) will prevent coupon holders from applying coupon at checkout.`
Screenshot 2023-12-29 at 12 53 49 PM

## Notifications
_NOTE: to get these notifications, you need to save a new coupon or update an existing coupon_

### Added Coupon
- `Added coupon "%s".`

Screenshot 2023-12-29 at 1 16 34 PM

### Updated Coupon
- `Updated coupon "%s".`

Screenshot 2023-12-29 at 1 17 37 PM

## Testing Instructions

* Stick the appropriate blog sticker to make coupons work on your memberships site. Hint: search wpcom for `is_coupons_and_gifts_enabled`. This allows the coupons to be used with the backend, otherwise the routes will not exist.
* Apply D133250-code to your sandbox. Sandbox the public API and subscribe.wordpress.com. Enable store sandbox.
* Navigate to `/earn/payments/` using one of the calypso live links below.
* Click "Add coupon" and add a coupon. Try various options. Click save.
* Edit the coupons by clicking the ellipsis menu and clicking edit. The same goes for deleting them.
* Create more coupons with different configurations and ensure they work as expected.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Review by millerf — APPROVED** [review]
WFM. IT also hides when the flag is not up.

One minor detail: it could have "Amount" title change to "Percentage" when switching... 

Nice work! 🚢 


Screenshot 2023-12-28 at 22 33 13

 
Screenshot 2023-12-28 at 22 32 57
Screenshot 2023-12-28 at 22 32 16

File: client/my-sites/earn/components/add-edit-coupon-modal/style.scss [hunk 1]
```diff
@@ -0,0 +1,148 @@
+@import "@wordpress/base-styles/breakpoints";
+@import "@wordpress/base-styles/mixins";
+@import "@automattic/components/src/styles/typography.scss";
+
+.memberships__dialog-sections {
+	display: grid;
+	grid-template: "item" 1fr / 1fr;
```
Comment (millerf) [inline client/my-sites/earn/components/add-edit-coupon-modal/style.scss:7]: TIL...

File: client/my-sites/earn/types.ts [hunk 2]
```diff
@@ -14,6 +14,25 @@ export type Product = {
 	tier?: number;
 };
 
+export type Coupon = {
```
Comment (millerf) [inline client/my-sites/earn/types.ts:17]: Nice touch

File: client/state/memberships/coupon-list/test/actions.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			expect( receiveDeleteCoupon( 123, 456 ) ).toEqual( {
+				couponId: 456,
+				siteId: 123,
+				type: MEMBERSHIPS_COUPON_DELETE,
+			} );
+		} );
+	} );
+
+	describe( 'requestAddCoupon', () => {
+		test( 'should return an action object with the new coupon', () => {
```
Comment (millerf) [inline client/state/memberships/coupon-list/test/actions.js]: Nice!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/10557200

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @chrismccluskey for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10557200).

**Pull request merged**